include/ensmallen_bits/callbacks/report.hpp

```diff
@@ -0,0 +1,558 @@
+   * @param objective Objective value of the current point.
+   */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void EndEpoch(OptimizerType& optimizer,
+                FunctionType& /* function */,
+                const MatType& /* coordinates */,
+                const size_t /* epoch */,
+                const double objective)
+  {
+    // In case StepTake() has been called first we clear the existing data.
+    if (!hasEndEpoch)
+    {
+      hasEndEpoch = true;
+
+      objectives.clear();
+      timings.clear();
+      gradientsNorm.clear();
+      stepsizes.clear();
+    }
+
+    objectives.push_back(objective);
+    timings.push_back(optimizationTimer.toc());
+
+    if (hasGradient)
+      gradientsNorm.push_back(gradientNorm);
+    
+    SaveStepSize(optimizer);
+  }
+
+  /**
+   * Callback function called once a step is taken.
+   *
+   * @param optimizer The optimizer used to update the function.
+   * @param function Function to optimize.
+   * @param coordinates Starting point.
+   * @param objective Objective value of the current point.
+   */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void StepTaken(OptimizerType& optimizer,
+                 FunctionType& /* function */,
+                 const MatType& /* coordinates */)
+  {
+    if (!hasEndEpoch)
+    {
+      objectives.push_back(objective);
+      timings.push_back(optimizationTimer.toc());
+
+      if (hasGradient)
+        gradientsNorm.push_back(gradientNorm);
+
+      SaveStepSize(optimizer);
+    }
+  }
+
+  /**
+   * Callback function called at any call to Evaluate().
+   *
+   * @param optimizer The optimizer used to update the function.
+   * @param function Function to optimize.
+   * @param coordinates Starting point.
+   * @param objectiveIn Objective value of the current point.
+   */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void Evaluate(OptimizerType& /* optimizer */,
+                FunctionType& /* function */,
+                const MatType& /* coordinates */,
+                const double objectiveIn)
+  {
+    objective = objectiveIn;
+    evaluateCalls++;
+  }
+
+  /**
+    * Callback function called at any call to EvaluateConstraint().
+    *
+    * @param optimizer The optimizer used to update the function.
+    * @param function Function to optimize.
+    * @param coordinates Starting point.
+    * @param constraint The index of the constraint;
+    * @param objective Objective value of the current point.
+    */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void EvaluateConstraint(OptimizerType& optimizer,
+                          FunctionType& function,
+                          const MatType& coordinates,
+                          const size_t /* constraint */,
+                          const double objective)
+  {
+    Evaluate(optimizer, function, coordinates, objective);
```

Comment:
Are you sure this is what we want here?  It looks like this will set `this->objective` to the objective given to `EvaluateConstraint()`, but maybe what we want to be doing instead is adding the `objective` input (which is the objective of the constraint) to the overall objective?  i.e.:

```
  objective += objectiveIn;
```

Correct me if I overlooked something there.

Reply:
Looks like this is handled now, so I'll resolve the conversation. :+1: